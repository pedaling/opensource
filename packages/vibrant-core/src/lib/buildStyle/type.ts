import type { CurrentTheme } from '@vibrant-ui/theme';

type StyleObject = {
  [property: string]: any;
};

export type BuildStyleFn = (styleObjects: StyleObject[], info: { theme?: CurrentTheme }) => Record<string, any>;
