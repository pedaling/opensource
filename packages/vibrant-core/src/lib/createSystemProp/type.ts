import type { Theme } from '@vibrant-ui/theme';
import type { Path } from '@vibrant-ui/utils';

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: Extract<Path<Theme>, 'colors' | 'opacity' | 'typography' | 'typographyWeight'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
