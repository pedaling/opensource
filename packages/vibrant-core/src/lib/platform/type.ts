export type Platform = 'android' | 'ios' | 'macos' | 'web' | 'windows';
