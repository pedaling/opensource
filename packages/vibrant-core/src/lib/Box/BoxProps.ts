import type { ComponentType, Ref } from 'react';
import type { DistributiveOmit, LayoutEvent } from '@vibrant-ui/utils';
import type { ReactElementChildren } from '../../types';
import { createInterpolation } from '../createInterpolation';
import { createShouldForwardProp } from '../createShouldForwardProp';
import { injectContext } from '../injectContext';
import type {
  BackgroundSystemProps,
  BorderSystemProps,
  ColorSystemProps,
  DisplaySystemProps,
  ElevationSystemProps,
  FlexboxSystemProps,
  InputSystemProps,
  InteractionSystemProps,
  OverflowSystemProps,
  PositionSystemProps,
  PseudoClassSystemProps,
  SizingSystemProps,
  SpacingSystemProps,
  SvgSystemProps,
  TextSystemProps,
  TransformSystemProps,
  TypographySystemProps,
} from '../props';
import {
  backgroundSystemProps,
  borderSystemProps,
  colorSystemProps,
  displaySystemProps,
  elevationSystemProps,
  flexboxSystemProps,
  inputSystemProps,
  interactionSystemProps,
  overflowSystemProps,
  positionSystemProps,
  pseudoClassSystemProps,
  sizingSystemProps,
  spacingSystemProps,
  svgSystemProps,
  textSystemProps,
  transformSystemProps,
  typographySystemProps,
} from '../props';

export type SystemProps = BackgroundSystemProps &
  BorderSystemProps &
  ColorSystemProps &
  DisplaySystemProps &
  ElevationSystemProps &
  FlexboxSystemProps &
  InputSystemProps &
  InteractionSystemProps &
  OverflowSystemProps &
  PositionSystemProps &
  PseudoClassSystemProps &
  SizingSystemProps &
  SpacingSystemProps &
  SvgSystemProps &
  TextSystemProps &
  TypographySystemProps &
  TransformSystemProps;

const systemProps = [
  ...backgroundSystemProps,
  ...borderSystemProps,
  ...colorSystemProps,
  ...displaySystemProps,
  ...elevationSystemProps,
  ...flexboxSystemProps,
  ...inputSystemProps,
  ...interactionSystemProps,
  ...overflowSystemProps,
  ...positionSystemProps,
  ...pseudoClassSystemProps,
  ...sizingSystemProps,
  ...spacingSystemProps,
  ...svgSystemProps,
  ...textSystemProps,
  ...typographySystemProps,
  ...transformSystemProps,
];

const systemPropNames = systemProps.filter(systemProp => !systemProp.disabled).map(systemProp => systemProp.propName);

export type BoxElements =
  | 'article'
  | 'aside'
  | 'button'
  | 'clipPath'
  | 'defs'
  | 'div'
  | 'footer'
  | 'g'
  | 'header'
  | 'img'
  | 'input'
  | 'label'
  | 'li'
  | 'linearGradient'
  | 'main'
  | 'mask'
  | 'nav'
  | 'path'
  | 'section'
  | 'span'
  | 'stop'
  | 'svg'
  | 'ul';

type ComponentProps<Component extends ComponentType | undefined> = Component extends undefined
  ? undefined
  : Component extends ComponentType<infer ComponentProps>
  ? ComponentProps
  : undefined;

export type BoxProps<
  BaseComponent extends ComponentType | undefined = undefined,
  ElementName extends BoxElements | undefined = undefined
> = SystemProps & {
  as?: ElementName;
  base?: BaseComponent;
  id?: string;
  onLayout?: (layoutEvent: LayoutEvent) => void;
} & DistributiveOmit<
    ComponentProps<BaseComponent> extends undefined
      ? (ElementName extends keyof JSX.IntrinsicElements
          ? JSX.IntrinsicElements[ElementName]
          : Record<never, never>) & {
          ref?: Ref<HTMLElement>;
          children?: ReactElementChildren;
        }
      : ComponentProps<BaseComponent>,
    keyof SystemProps
  >;

export type { LayoutEvent };

export const interpolation = injectContext(
  createInterpolation(systemProps, { display: 'flex', boxSizing: 'border-box', position: 'relative' })
);

export const shouldForwardProp = createShouldForwardProp(systemPropNames);
