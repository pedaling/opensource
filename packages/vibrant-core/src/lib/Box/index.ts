export type { BoxProps, LayoutEvent } from './BoxProps';
export { Box } from './Box';
