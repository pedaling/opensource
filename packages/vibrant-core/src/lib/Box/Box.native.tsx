import type { ComponentType } from 'react';
import { forwardRef, memo } from 'react';
import type { LayoutChangeEvent } from 'react-native';
import { StyleSheet, TextInput, TouchableOpacity, View } from 'react-native';
import styled from '@emotion/native';
import { isDefined } from '@vibrant-ui/utils';
import { getElementPosition } from '../getElementPosition';
import { OnColorContainer } from '../OnColorContainer';
import type { BoxProps } from './BoxProps';
import { interpolation, shouldForwardProp } from './BoxProps';

const transformAs = (as: keyof JSX.IntrinsicElements): ComponentType => {
  switch (as) {
    case 'button':
      return TouchableOpacity;
    case 'input':
      return TextInput;
    default:
      return View;
  }
};

export const Box = memo(
  styled(
    forwardRef<HTMLDivElement, BoxProps & { style: any }>(
      (
        {
          as,
          base,
          style,
          onLayout,
          backgroundColor,
          id,
          role,
          ariaLabel,
          ariaLabelledBy,
          ariaChecked,
          ariaSelected,
          ariaCurrent: _ariaCurrent,
          ...restProps
        },
        ref
      ) => {
        const { BaseComponent, props, ...restStyle } = StyleSheet.flatten(style);

        const Component = BaseComponent ?? base ?? transformAs(as ?? 'div');

        const handleLayout = onLayout
          ? async (event: LayoutChangeEvent) => {
              const layout = await getElementPosition(event.currentTarget);

              onLayout(layout);
            }
          : undefined;

        const element = (
          <Component
            ref={ref}
            id={id}
            role={role}
            style={restStyle}
            onLayout={handleLayout}
            accessibilityLabel={ariaLabel}
            accessibilityLabelledBy={ariaLabelledBy}
            {...(isDefined(ariaChecked) || isDefined(ariaSelected)
              ? { accessibilityState: { checked: ariaChecked, selected: ariaSelected } }
              : {})}
            collapsable={ref ? false : undefined}
            {...(base ? { as } : {})}
            {...restProps}
            {...props}
          />
        );

        if (backgroundColor) {
          return <OnColorContainer backgroundColor={backgroundColor}>{element}</OnColorContainer>;
        }

        return element;
      }
    ),
    {
      shouldForwardProp,
    }
  )(interpolation)
);

Box.displayName = 'Box';
