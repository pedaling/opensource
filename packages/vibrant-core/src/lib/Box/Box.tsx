import type { ComponentType, FC, ReactElement } from 'react';
import { createElement, forwardRef, useCallback } from 'react';
import styled from '@emotion/styled';
import { useLayoutEvent } from '@vibrant-ui/utils';
import { OnColorContainer } from '../OnColorContainer';
import type { BoxElements, BoxProps } from './BoxProps';
import { interpolation, shouldForwardProp } from './BoxProps';

export const Box = styled(
  forwardRef<HTMLDivElement, BoxProps>(({ as, base, onLayout, ariaLabel, backgroundColor, ...restProps }, ref) => {
    const Component = base ? (base as ComponentType<any>) : undefined;

    const { ref: layoutEventRef } = useLayoutEvent(onLayout);

    const composeRef = useCallback(
      (node: HTMLDivElement) => {
        if (ref) {
          if (typeof ref === 'function') {
            ref(node);
          } else {
            ref.current = node;
          }
        }

        layoutEventRef(node);
      },
      [ref, layoutEventRef]
    );

    if (Component) {
      return (
        <OnColorContainer backgroundColor={backgroundColor}>
          <Component ref={composeRef} {...(as ? { as } : {})} aria-label={ariaLabel} {...restProps} />
        </OnColorContainer>
      );
    }

    return (
      <OnColorContainer backgroundColor={backgroundColor}>
        {createElement(as ?? 'div', {
          ref: composeRef,
          'aria-label': ariaLabel,
          ...restProps,
        })}
      </OnColorContainer>
    );
  }),
  {
    shouldForwardProp,
  }
)(interpolation) as <
  BaseComponent extends ComponentType<any> | undefined = undefined,
  ElementName extends BoxElements | undefined = undefined
>(
  props: BoxProps<BaseComponent, ElementName>
) => ReactElement;

(Box as FC).displayName = 'Box';
