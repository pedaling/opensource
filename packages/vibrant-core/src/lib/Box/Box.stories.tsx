import type { ComponentStory, ComponentMeta } from '@storybook/react';
import { Box } from './Box';

export default {
  title: 'Box',
  component: Box,
  args: {
    width: 300,
    height: 300,
    backgroundColor: 'primary',
  },
} as ComponentMeta<typeof Box>;

export const Basic: ComponentStory<typeof Box> = props => <Box {...props}>Box</Box>;
