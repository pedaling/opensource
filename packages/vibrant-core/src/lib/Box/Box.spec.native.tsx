import { View, Text, Pressable } from 'react-native';
import type { ReactNativeRenderer } from '@vibrant-ui/utils';
import { createReactNativeRenderer } from '@vibrant-ui/utils';
import { Box } from './Box';

describe('Native <Box />', () => {
  const { render } = createReactNativeRenderer();

  let renderer: ReactNativeRenderer;

  describe('as prop', () => {
    describe('when empty box created', () => {
      beforeEach(() => {
        renderer = render(<Box />);
      });

      it('View component used', () => {
        expect(renderer.container.findByType(View)).toBeTruthy();
      });

      it('match snapshot', () => {
        expect(renderer.toJSON()).toMatchSnapshot();
      });
    });

    describe('when span box created', () => {
      beforeEach(() => {
        renderer = render(<Box as="span" />);
      });

      it('Text component used', () => {
        expect(renderer.container.findByType(Text)).toBeTruthy();
      });

      it('match snapshot', () => {
        expect(renderer.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('base prop', () => {
    describe('when Pressable box created', () => {
      beforeEach(() => {
        renderer = render(<Box base={Pressable} id="component" />);
      });

      it('Pressable element created', () => {
        expect(renderer.container.findByType(Pressable)).toBeTruthy();
      });

      it('id prop injected', () => {
        expect(renderer.container.findByType(Pressable).props['id']).toBe('component');
      });

      it('match snapshot', () => {
        expect(renderer.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('default style', () => {
    describe('when styled box created', () => {
      beforeEach(() => {
        renderer = render(<Box width={300} height={300} />);
      });

      it('style apply on View', () => {
        expect(renderer.container.findByType(View)).toHaveStyle({
          width: 300,
          height: 300,
        });
      });

      it('match snapshot', () => {
        expect(renderer.toJSON()).toMatchSnapshot();
      });
    });
  });
});
