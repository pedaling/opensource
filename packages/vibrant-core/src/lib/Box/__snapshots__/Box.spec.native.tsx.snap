// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Native <Box /> as prop when br box created match snapshot 1`] = `
<Text
  style={
    Object {
      "alignContent": "stretch",
      "display": "flex",
      "flexDirection": "column",
      "flexShrink": 0,
    }
  }
>
  

</Text>
`;

exports[`Native <Box /> as prop when empty box created match snapshot 1`] = `
<View
  style={
    Object {
      "alignContent": "stretch",
      "display": "flex",
      "flexDirection": "column",
      "flexShrink": 0,
    }
  }
/>
`;

exports[`Native <Box /> as prop when span box created match snapshot 1`] = `
<Text
  lineBreakMode="tail"
  style={
    Object {
      "alignContent": "stretch",
      "display": "flex",
      "flexDirection": "column",
      "flexShrink": 0,
    }
  }
/>
`;

exports[`Native <Box /> base prop when Pressable box created match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  id="component"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignContent": "stretch",
      "display": "flex",
      "flexDirection": "column",
      "flexShrink": 0,
    }
  }
/>
`;

exports[`Native <Box /> default style when styled box created match snapshot 1`] = `
<View
  style={
    Object {
      "alignContent": "stretch",
      "display": "flex",
      "flexDirection": "column",
      "flexShrink": 0,
      "height": 300,
      "width": 300,
    }
  }
/>
`;
